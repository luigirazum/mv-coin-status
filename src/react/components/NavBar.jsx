import { useDispatch, useSelector } from 'react-redux';
import { NavLink, useParams } from 'react-router-dom';
import { selectFilterBy } from '../../redux/coins/coinsSelectors';
import { selectCoinDetailsById, selectDetailsError } from '../../redux/details/detailsSelectors';
import { clearFilterCoins, filterCoins } from '../../redux/coins/coinsSlice';
import {
  CancelIcon,
  CoinIcon, CoinsIcon, GoBackIcon, HomeIcon, LensIcon,
} from './icons';

const NavBar = () => {
  const { id } = useParams();
  const dispatch = useDispatch();
  const coin = useSelector((store) => selectCoinDetailsById(store, id));
  const detailsError = useSelector(selectDetailsError);
  const filterBy = useSelector(selectFilterBy);

  const onChangeHandler = (e) => {
    const filter = e.target.value.trim().toLowerCase();
    if (filter) {
      dispatch(filterCoins(filter));
    } else {
      dispatch(clearFilterCoins());
    }
  };

  const onClickHandler = (e) => {
    const { id: btnId } = e.target;
    if (btnId.includes('cancel')) dispatch(clearFilterCoins());
  };

  const onSubmitHandler = (e) => {
    e.preventDefault();
  };

  return (
    <header>
      <nav>
        {(!coin && !detailsError) ? (
          <>
            <NavLink to="/" end className="fsControls navLink">
              <HomeIcon />
              <span className="fsControls navName">Home</span>
            </NavLink>
            <NavLink to="." end className="fsControls navLink" id="cancel-filter" onClick={onClickHandler}>
              <CoinsIcon />
              <span className="fsControls navName">Coins</span>
            </NavLink>
          </>
        ) : (
          <>
            <NavLink to=".." relative="path" end className="fsControls navLink">
              <GoBackIcon />
            </NavLink>
            <NavLink to={id} end className="fsControls navLink navCoin">
              <CoinIcon />
              <span className="fsControls navName">{coin ? coin.name : `${id}coin`}</span>
            </NavLink>
          </>
        )}
      </nav>
      {(!(coin && id) && !(id && !coin)) && (
      <form id="filter" onSubmit={onSubmitHandler}>
        <input
          type="text"
          id="field"
          placeholder="filter by name/symbol"
          value={filterBy}
          onChange={onChangeHandler}
        />
        <button
          type="button"
          id="cancel"
          data-testid={`${filterBy ? 'cancel-btn' : 'filter-btn'}`}
          className={`ctrl-btn ${filterBy ? 'cancel-btn' : 'filter-btn'}`}
          onClick={onClickHandler}
        >
          {filterBy ? <CancelIcon /> : <LensIcon />}
        </button>
      </form>
      )}
    </header>
  );
};

export default NavBar;
