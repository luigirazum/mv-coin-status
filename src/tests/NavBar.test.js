import { create } from 'react-test-renderer';
import renderWithProvider from '../react/utils/renderUtils';
import { mockCoins, mockBitcoin } from './mocks/mockData';
import NavBar from '../react/components/NavBar';

const testState = {
  coins: {
    assets: [...mockCoins.coins],
    status: { type: 'idle' },
    filter: { active: false, by: '' },
    error: null,
  },
};

const testDetailsBitcoin = {
  details: {
    asset: mockBitcoin,
    status: { type: 'idle' },
    error: null,
  },
};

const testRouter = {
  routerWrapper: {
    memory: true,
  },
  routes: ['/coins'],
  path: '/coins',
};

let testRender = null;
let container = null;
let queryByText;
let queryByPlaceholderText;
let queryByTitle;

const resetTestEnv = () => {
  container.remove();
  container = null;
  testRender = null;
};

beforeEach(() => {
  // setup a DOM element as a render target
  testRender = renderWithProvider(
    <NavBar />,
    { preloadedState: { ...testState } },
    { ...testRouter },
  );
  ({ container } = testRender);
  ({ queryByText, queryByPlaceholderText } = testRender);
  document.body.appendChild(container);
});

afterEach(() => {
  //  cleanup on exiting
  resetTestEnv();
});

/** START: NavBar component test */
describe('NavBar component tests', () => {
  /** START: NavBar render at /coins */
  describe('at /coins renders a NavBar with Home and Coins links', () => {
    test('it has a Home link', () => {
      expect(queryByText('Home')).toBeInTheDocument();
    });
    test('it has a Coins link', () => {
      expect(queryByText('Coins')).toBeInTheDocument();
    });
    test('it has a Filter field', () => {
      expect(queryByPlaceholderText(/filter by/i)).toBeInTheDocument();
    });
  });
  /** END: NavBar render at /coins */

  /** START: NavBar render at /coins/:coinId */
  describe('at /coins/:id renders a NavBar with < and coinName links', () => {
    beforeEach(() => {
      resetTestEnv();

      const testDetailsState = {
        ...testState,
        ...testDetailsBitcoin,
      };
      // setup a DOM element as a render target
      testRender = renderWithProvider(
        <NavBar />,
        { preloadedState: { ...testDetailsState } },
        {
          ...testRouter,
          routerWrapper: { memory: true },
          routes: ['/coins/bitcoin'],
          path: '/coins/:id',
        },
      );
      ({ container } = testRender);
      ({ queryByText, queryByPlaceholderText, queryByTitle } = testRender);
      document.body.appendChild(container);
    });

    test('it has a Home link', () => {
      expect(queryByTitle('back-icon')).toBeInTheDocument();
    });
    test('it has a Bitcoin link', () => {
      expect(queryByText(/Bitcoin/i)).toBeInTheDocument();
    });
    test('it does\'t have a Filter field', () => {
      expect(queryByPlaceholderText(/filter by/i)).toBeNull();
    });
  });
  /** END: NavBar render at /coins/coinId */

  /** START: NavBar snapshot test */
  describe('NavBar component snapshot test', () => {
    test('snapshot renders NavBar correctly', () => {
      const { component } = testRender;

      const tree = create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  /** END: NavBar snapshot test */
});
/** END: NavBar component test */
