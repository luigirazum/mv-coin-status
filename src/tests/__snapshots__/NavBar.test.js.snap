// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar component tests NavBar component snapshot test snapshot renders NavBar correctly 1`] = `
<header>
  <nav>
    <a
      className="fsControls navLink"
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
    <a
      aria-current="page"
      className="fsControls navLink active"
      href="/coins"
      onClick={[Function]}
      onClickCapture={[Function]}
    >
      Coins
    </a>
  </nav>
  <form
    id="filter"
    onSubmit={[Function]}
  >
    <input
      id="field"
      onChange={[Function]}
      placeholder="filter by name/symbol"
      type="text"
      value=""
    />
    <button
      className="ctrl-btn"
      id="reset"
      onClick={[Function]}
      type="button"
    >
      X
    </button>
  </form>
</header>
`;
